pkgs/pkgs-lib/formats.nix

```diff
@@ -131,8 +141,20 @@ rec {
               if lib.isList val then listToValue val else val
             )) value
           else value;
-      in pkgs.writeText name (lib.generators.toINI (removeAttrs args ["listToValue"]) transformedValue);
 
+        filterSections = keepAttrs: lib.filterAttrs (_: val: lib.isAttrs val == keepAttrs) transformedValue;
+
+        genArgs = (removeAttrs args ["globalSection" "listToValue"]);
+      in
+        if globalSection then
+          lib.generators.toINIWithGlobalSection genArgs {
+            globalSection = filterSections false;
+            sections = filterSections true;
```

Comment:
This is problematic for the same reasons discussed in https://github.com/NixOS/nixpkgs/issues/107491, which ultimately [led](https://github.com/NixOS/nixpkgs/pull/118925) to [`toINIWithGlobalSection`](https://github.com/NixOS/nixpkgs/pull/164088).

`pkgs.formats` should not add its own interface on top of that decision, but rather provide the same interface, which is apparently done in https://github.com/NixOS/nixpkgs/pull/201686.

Reply:
I hadn't thought of that limitation, thanks for the links.

I do think forcing the NixOS config to split the global and other sections would not be a nice user experience and want to avoid that.  
What do you think of having:
- `formats.ini`: `withGlobalSection` that splits `value` into `globalSection` and `sections` itself (opt-in of course)
- `formats.iniWithGlobalSection` that takes an attr with explicit `globalSection` and `section`, just like `generators.toINIWithGlobalSection`

`ini` would be the most common use case: you can't represent all potential configs, but anyways most software doesn't need to represent that edge case, so it's fine.  
`iniWithGlobalSection` would be used when you actually need to support that edge case.

If you don't think that's a good idea, then I can always implement the `settings` splitting in the `gitea` module (would also need to change the `settings` type to not rely on `(formats.ini {}).type`).
I think it'd be nice not to have to do that in every module, but up to you, I just want to merge my changes to Gitea 😄

Reply:
I don't think it's a good idea to add custom interfaces on top. `pkgs.formats` should ideally just be an extended `lib.generators`. Either we add this splitting idea to `lib.generators` and `pkgs.formats` or neither. Now having the interface for `lib.generators.toINIWithGlobalSection` already decided, it makes most sense to just use the same interface for `pkgs.formats` too.

However I do agree that in a most cases, this splitting idea would enable a much nicer interface. I think it might be okay to have that _if_ we can make sure that we won't run into a situation where we can't represent a configuration. 

This essentially means: Check all the potential options of a whole bunch of programs using INI config files to see if you can find one that uses the same name for a global variable and a section. Maybe there's also an example where global variables are dynamic and can be freely picked, same for section names. You could also automate this by [looking at all INI files on GitHub](https://github.com/search?l=&p=3&q=language%3AINI&type=Code) to see if there's a clash. You might also want to check INI libraries in other programming languages to see how they handle global sections.

We need to be really sure because we'd essentially be redefining the semantics of global sections in the INI format. If we run into an example where NixOS can't represent some configuration in the future, we'd have to give pressure for the authors to change the config names to not clash.

Reply:
Put this on the back-burner for a while, and looking into it again now.

> Either we add this splitting idea to lib.generators and pkgs.formats or neither. Now having the interface for lib.generators.toINIWithGlobalSection already decided, it makes most sense to just use the same interface for pkgs.formats too.

I agree with this, but I'd prefer to make backwards compatible changes to the ini formats and generators so `formats.ini` can support splitting.  
Of course `ini` wouldn't support the conflicting global and section names edge case, but the key would be to make switching a module's format from `ini` to (a not yet created) `iniWithGlobalSection` transparent for the user's `settings`.

What I'm thinking is slightly changing  `toINIWithGlobalSection`'s semantics by changing the params  
from `{ globalSection, sections }` to `{ extraGlobals, values }`.  
`extraGlobals` would only accept global values (`attrsOf iniAtom`), as `globalSection` does, but it's intended use would mainly be for values whose name conflicts with a section.
`values` would be loosened compared to `sections` to accept both global values and sections (`either iniAtom (attrsOf iniAtom)`), and split into globals and sections.
Pseudo code to express the relationship between old and new params:  
- `globalSection = extractGlobals values // extraGlobals`
- `sections = extractSections values`

We'd have an assert to ensure `intersectAttrs (extractGlobals values) extraGlobals == { }`, so that nothing gets accidentally shadowed during the attrset merge.

This would make it possible to define `ini` as `iniWithGlobalSection` with `extraGlobals = { }`. Providing an upgrade path for any module that would someday need to support _the edge case_: switch the format from `ini` to `iniWithGlobalSection` and expose a second free-form attrset in the options that would feed be used as `extraGlobals`.  
Maybe something like so (this is just an example, and I'm not proposing a design for the module API):
```nix
{
  settings = { conflictingName = { ... }; };
  settingsExtraGlobals = { conflictingName = 0; };
}
```

How does that sound?

Reply:
Hmm it's an interesting idea, but I think we can make it work without pushing the edge case responsibility to all the modules. To do this we just need to detect whether there's a `global` keyword, and if so, request the user to distinguish between global sections and normal ones. More formally:
- Each definition of a `types.ini` option can either be of type `attrsOf (attrsOf iniAtom)` or `{ globalSection :: attrsOf iniAtom, sections :: attrsOf (attrsOf iniAtom) }`. In case of ambiguity, an appropriate error is thrown. Additionally if the type would be unambiguously the former one, but `globalSection` is a section name, an appropriate error is thrown. Internally the former gets coerced to the latter.

Here's some examples:
```nix
{
  settings = lib.mkMerge [
    # Valid, unambiguously the former type
    {
      foo.bar = 10;
      bar.baz = 20;
    }
    # Valid, unambiguously the latter type
    {
      globalSection.foo = 10;
      sections.bar.baz = 20;
    }
    # Invalid, ambiguous type
    {
      globalSection.foo = 20;
    }
    # Valid, only a global section, arguably the main drawback
    {
      globalSection.foo = 20;
      sections = {};
    }
  ];
}
```